### PR TITLE
Use --no-implicit-optional for type checking

### PR DESCRIPTION
**Pull request by hauntsaninja**
## Change Summary

This makes type checking PEP 484 compliant (as of 2018).
mypy will change its defaults soon.

See:
https://github.com/python/mypy/issues/9091
https://github.com/python/mypy/pull/13401

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/pydantic/pydantic/blob/master/changes/README.md) for details)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by hauntsaninja** [comment]
please review :-)

File: changes/4378-hauntsaninja.md [hunk 1]
```diff
@@ -0,0 +1 @@
+Use ``--no-implicit-optional`` for type checking.
```
Comment (samuelcolvin) [inline changes/4378-hauntsaninja.md:1]: This is markdown, not rst.

**Review by samuelcolvin — COMMENTED** [review]
Otherwise looks good, thank you.

**Comment by hauntsaninja** [comment]
I fixed the markdown, but it looks like tests on Python 3.8 are failing with a weird Literal[None] issue. I see the code to work around this — was that issue flaky or has something about my change reactivated it?

**Comment by samuelcolvin** [comment]
That's very weird, but I haven't seen that test being flakey before. I think it much be caused by your change, or be new.

**Comment by samuelcolvin** [comment]
I don't see it on other newer builds, so it must be caused by your changes.

**Comment by hauntsaninja** [comment]
I'll take a detailed look at it, but if I can't figure it out, how would you feel about potentially using the 3.7 code for 3.8 as well? If it can be activated by this seemingly no-op change, there may be a real issue here.

https://github.com/hauntsaninja/pydantic/blob/dab478aa81a59e0a5c69a0fe57bfe34f233994ff/pydantic/typing.py#L319

**Comment by samuelcolvin** [comment]
It looks like we have a custom implementation for a reason, but if it's really not necessary, we could remove it.

**Comment by samuelcolvin** [comment]
### Notice

See [twitter :bird:](https://twitter.com/pydantic/status/1560554679473303552), I've you'd like this to be included in V1.10, please fix it and request a review **TODAY**.

Or if you need this in V1.10 but don't have time to complete it (or aren't the author), please comment here and on #4324

**Comment by samuelcolvin** [comment]
Let's wait and adopt this in the V2 codebase.

**Comment by Fokko** [comment]
Great to see this happening! 👍🏻 Don't forget to add `py.typed`: https://peps.python.org/pep-0561/#packaging-type-information

**Comment by samuelcolvin** [comment]
We've had a `py.typed` file since 2017?

**Comment by Fokko** [comment]
@samuelcolvin ah missed that one. I expected it to be on the root level. 👍🏻

**Comment by samuelcolvin** [comment]
The code base will change too much for V2 to keep this pr, but I'm tracking the issue in #4424

**Pull request closed without merging**